### DISCUSSION TITLE
Telegraf.js Scenes. Help please.

### DISCUSSION THREAD

**Discussion by Alheimurine** [telegraf/telegraf#1907] [Q&A] [answered]
Hello! Anyone help me what is my error?
Bot have 7 scenes. 
They open up with select on inline.keyboard. 
If i open 1st scene, and dont complete her, and open 2nd scene - overlap occurs. 
1st scene - buy car / 2nd scene - sell car 
The point this : 
- select buy button -> select car make -> select from suggested cities
If i select buy button (buy scene) -> do not complete this scene and straight select sell button (2nd scene), that will will be a transition to next step in 1st scene
Handler installed - 'callback_query'
My code: 
- Menu
```
bot.action('menu', async (ctx)=> {
    await ctx.replyWithHTML('<b>В меню ниже выберите вопрос, который вас интересует.</b>', Markup.inlineKeyboard(
        [
            [Markup.button.callback('buy','buy'), Markup.button.callback('sell','sell')],
            ...and other 5 buttons
        ]
    ))
})
```
- Buy 
```
const buyOpen = new Composer()
buyOpen.action('buy', async (ctx) => {
    try {
        if (!ctx.session) {
            ctx.session = {}
        }
        await ctx.reply('Выберите:',Markup.inlineKeyboard(
            [
                [Markup.button.callback('LADA', 'LADA')],
                [Markup.button.callback('CHERY', 'CHERY')],
                [Markup.button.callback('menu', 'menu')]
            ]
        ))
        return ctx.wizard.next()
    } catch(e) {
        console.error(e)
    }
})
const markBuy = new Composer()
markBuy.on('callback_query', async (ctx) => {
    try {
        ctx.session.buyCarInfo = {}
        ctx.session.buyCarInfo.markBuy = ctx.update.callback_query.data
        if (ctx.update.callback_query.data === 'LADA') {
            await ctx.reply('Select:', Markup.inlineKeyboard(
                [
                    [Markup.button.callback('city 1','city 1')],[Markup.button.callback('city 2','city 3')],
                    ... and other cities
                ]
            ))
        } else if (ctx.update.callback_query.data === 'CHERY') {
            await ctx.reply('Select:', Markup.inlineKeyboard(
                [
                    [Markup.button.callback('city 1','city 1')]
                ]
            ))
        } else {
            await ctx.reply('Select:', Markup.inlineKeyboard(
                [
                    [Markup.button.callback('city1','city 1')],[Markup.button.callback('city 2','city 2')],
                ]
            ))
        }
        return ctx.wizard.next()  
    } catch(e) {
        console.error(e)
    }
})

const city = new Composer()
city.on('callback_query', async (ctx) => {
    try {
        const sort = (mark, city, buttons) => {
            if (ctx.session.buyCarInfo.markBuy === mark && ctx.update.callback_query.data === city) {
                ctx.replyWithHTML(`Select:`, 
                    Markup.inlineKeyboard(
                        [
                            ...buttons.map(([name, link]) => [Markup.button.url(name, link)]),
                            [Markup.button.callback('Menu', 'menu')]
                        ]  
                    ),
                )
            }
        }
        if (ctx.session.buyCarInfo.city !== ctx.update.callback_query.data) {
            processing depending on select
            return ctx.scene.leave()
        } 
    } catch(e) {
        console.error(e)
    }
})

const buy = new Scenes.WizardScene('buyScene', buyOpen, markBuy, city)

```
- Sell 
similar to buying, but with its own logic
- Calls 
```
bot.action('buy', async (ctx) => {
    await ctx.scene.enter('buyScene')
})
bot.action('sell', async (ctx) => {
    await ctx.scene.enter('sellScene')
})
```
Im tried to solve in this way: 
```
bot.action('buy', async (ctx) => {
    await ctx.scene.leave() / and ctx.scene.current.... 
    await ctx.scene.enter('buyScene')
})
```
But this way requiers double click on the switch button to reenter scene
Help please. Me need switch between scenes without completing them

**Comment by MKRhere** [comment]
Right, so the issue is that you're listening on `callback_query` inside the scene, and you trigger an unrelated `callback_query` (via pressing an old button), and that calls your next step. What do you want to happen when user presses "sell" while in the buy scene?

**Reply by Alheimurine** [reply]
We need the ability to freely vote on the script, even if they are not finished

**Reply by MKRhere** [reply]
You want all buttons to be interactive regardless of which scene they are in?

1. If I press "buy", I get the buy menu.
2. Then I go back and press "sell", I'll get the sell menu.
3. If I go back to the old buy menu and press a button, it should work?

**Reply by MKRhere** [reply, marked as answer]
If you don't need step 3, the solution is simple. Move your top-level actions under stage like so:

```TS
const stage = new Stage([ buy, sell ]);
stage.action("buy", ctx => ctx.scene.enter("buyScene"));
stage.action("sell", ctx => ctx.scene.enter("sellScene"));
```

These will work identically to `bot.action`, with the exception that they intercept events before they enter scenes, and are still able to use `ctx.scene`. The downside is that buttons in buyScene will not work while user is in sellScene. At that point, you must wonder if the inline keyboard needs to be inside scenes at all.

**Reply by Alheimurine** [reply]
Yes. Me need independent action. If user press buy , but he dont finished this scene(error press or other reason) and he press other menu buttons - all do work. I would like to simply press the menu button, respectively, when changing a scene, so that the old one closes without saving progress, and the new one opens from the very beginning. 
Sorry for my bad english. I partially use a translator.

**Reply by MKRhere** [reply]
The other approach is to just move all the keyboards and their callback listeners outside to global (bot) scope, and the scenes just handle second step onwards. That way all actions are available all the time.

**Reply by Alheimurine** [reply]
hm. 
take it outside?

**Reply by MKRhere** [reply]
Actually your previous comment
 
> close without saving progress, and the new one opens from the very beginning

suggests you don't want inline keyboard from buy to be active anymore if user presses the "sell" button. In that case you can just follow the code snippet I posted above.

**Reply by Alheimurine** [reply]
Okay! Thanks, ill try it. If it works how me need, I'll let you know

**Reply by Alheimurine** [reply]
Friend, this solution seems to have helped me! Thanks!
